### ISSUE TITLE
Support mixture of guest and host robust futexes

### ISSUE THREAD

**Issue by Sonicadvance1** [FEX-Emu/FEX#809]
We can't safely modify the guest's robust list 
  - Memory might not be sufficient, or it will just overwrite our additions on change
We also can't safely consume the full guest robust and append it to our own
  - We don't have a full view of glibc's robust list
  - Might work if we capture all HOST get/set lists and modifying it? Needs seccomp filters then.
  - Would require us to track our own list, append the glibc head to our tail, and append the guest one to that tail as well.
  - Same problem where host glibc could potentially overwrite the and end that contains the guest.
  - Can't guarantee no host robust-futex usage due to thunks.
 We have no interface in to setting 32bit robust list
  - Needs the syscall32 API

Proposal to the kernel?
- New robust list indexing API?
- Something that uses the key api?
- long set_robust_list_key(key_t key, struct robust_list_head *head, size_t len, int listflg);
  - IPC_CREAT - Create a new robust list associated with key
  - IPC_EXCL - same to ensure the syscall fails if already created
- long get_robust_list_key(key_t key, int pid, struct robust_list_head **head_ptr,
                            size_t *len_ptr);
- long del_robust_list_key(key_t key);
  - Disassociates robust list of key from process and deletes?

Allows us to create multiple robust lists tied to a pid.
Still have the regular get/set robust list path, just additional list arrays on top of that.
Expectation is to only have a handful of robust lists, can be a linear scan lookup.
Expose the 32bit path through the "regular" 32bit syscall path.

I'm already dreading bringing this up on the mailing list.

**Comment by Amanieu** [comment]
I don't think this is worth the effort to support. Very little software actually relies on robust pthread mutexes. And even then you'll only notice if a process crashes while holding a mutex in a shared memory region.

**Comment by phire** [comment]
I see potential for various userspace workarounds, so I get the feeling the kernel mailing list won't be open to such ideas. 

The problem is most of those workarounds require modifications to guest (and maybe host) glibc. Maybe upstream glibc would be open to a clean patch that fixes this solution, I'm thinking:

 * An api which allows a process to add arbitrary raw futexes to the robust list
 * A separate api which allows hooking a callback on internal robust futex creation/deletion

FEX could then register a thunked callback that takes robust futexes from the guest glibc and inserts them into the host glibc's robust list. 

------ 

@Amanieu, you are hopefully right. This issue is just super low priority and might never get fixed.
